_posts/2020-03-16-dask-usage-at-blue-yonder.markdown

```diff
@@ -0,0 +1,161 @@
+A source dataset is read, computations are performed on the data, and the result is written to
+a target dataset.
+In many cases, the layout of the source dataset (the partitioning, i.e., what data resides in which blob)
+is used for parallelization. This means the algorithms work independently on the individual
+blobs of the source dataset. Therefore, our respective Dask graphs again look very simple, with parallel nodes each performing
+the sequential operations of reading in the data from a source dataset blob, doing some computation on it,
+and writing it out to a target dataset blob. Again, there is a final reduction node writing the target
+dataset metadata. We typically use Dask's Delayed interface for these computations.
+
+![Simple Dask graph with parallel nodes and final reduction](/assets/images/2020-03-16-dask-usage-at-by-graph.png)
+
+### Dataset maintenance
+
+We use Dask/Distributed for performing maintenance operations like garbage collection on datasets.
+Garbage collection is necessary since we use a lazy approach when updating datasets and only delete or update
+references to blobs in the dataset metadata, deferring the deletion of no longer used blobs.
+In this case, Dask is used to parallelize the delete requests to the storage provider (Azure blob service).
+In a similar fashion, we use Dask to parallelize copy operations for backing up datasets.
+
+### Repartitioning a dataset
+
+Our algorithms rely on source datasets that are partitioned in a particular way. It is not always
+the case that we have the data available in a suitable partitioning, for instance, if the data results
+from a computation that used a different partitioning. In this case, we have to repartition the data.
+This works by reading the dataset as a [Dask dataframe](https://docs.dask.org/en/latest/dataframe.html) 
+repartitioning the dataframe using network shuffle, and writing it out again to a dataset.
+
+## Ad hoc data analysis
+
+Finally, our data scientists also use Dask for ad hoc analysis of data that is too large to
+fit into an individual compute node's memory.
+
+## Dask cluster setup at Blue Yonder
+
+We run a somewhat unique setup of Dask clusters that is partly driven by the specific requirements
+of our domain and partly historically grown. We are running a large number of individual clusters
+that are homogeneous in size with many of the clusters dynamically scaled. The peculiarities of this
+setup have in some cases triggered edge cases and uncovered bugs, which lead us to submit a number
+of upstream pull requests.
+
+### Cluster separation
+
+We are running separate Dask clusters for separate customers in order to isolate the customers from each other.
+For the same reason, we also run separate clusters for production, staging, and development environments. 
+
+But it does not stop there. Our compute pipeline involves different steps, for instance, downloading the input data
```

Comment:
```suggestion
But it does not stop there. Our compute pipeline involves different steps, for instance downloading the input data
```

unless there is some comma rule that I don't know about ;-)

Reply:
This section describes what I referred to as environment isolation.
I would motivate it with having different products which build onto each others and are running sequentially. Since every product / application operates mostly autonomously, we separated the environments off each others. We might also drop a note about having different runtime environnments (python only, C++, lately even java, etc.)

Reply:
But that's the **rest of this paragraph**, that you're talking about.
This sentence/part really focuses on what the pipeline is comprised of and therefore is not really Cluster separation?!


So IMHO, this sections does not convey a clear message and as a result there is a disconnect between heading and content.

Reply:
I wanted to motivate why we are using different clusters for different steps of the pipeline; I think that's what's special about us. I'll make this more clear.